src/blog/2023/04/chatgpt-nodered-fcn-node.md

```diff
@@ -0,0 +1,64 @@
+---
+title: Chat GPT in Node-RED Function Nodes
+subtitle: New Node-RED function with embedded ChatGPT is now open-sourced!
+description: Save yourself time when working on Node-RED with these three tips.
+date: 2023-04-28 12:00:00
+authors: ["joe-pavitt", "stephen-mclaughlin"]
+image: /images/blog/tile-chatgpt-fcn-node.jpg
+---
+
+Recently we [posted a demo of ChatGPT integration into a duplicate of the Node-RED function node](https://www.linkedin.com/posts/flowforge_chatgpt-with-node-red-function-nodes-activity-7052725869684953088-2yOA?utm_source=share&utm_medium=member_desktop) 
+onto our social media accounts. We have now [open-sourced](https://github.com/flowforge/node-red-function-gpt) this for all to play with, and **welcome any and all contributions**.
+<!--more-->
+
+## How it Works
+
+OpenAI make a collection of their [Generative AI models](https://platform.openai.com/docs/models) available
+via an API. We are wrapping OpenAI's [node.js module](https://www.npmjs.com/package/openai), and in particular
+using the `openai.createChatCompletion()` functionality. For this API, you provide a chat history, and ChatGPT will
+respond with the next entry in that conversation.
+
+In order to "train" ChatGPT for our use case of populating Node-RED function nodes, we first tried a collection of prompts, defining specific
+requirements for the contents, e.g. _"Always write Javascript"_, _"Never include the wrapping function definition"_,
+_"Assume the input is always msg"_.
+
+It turns out though, that we were over-engineering it, we were not getting reliable results and ended up
+realising that ChatGPT's existing knowledge of Node-RED was sufficient such that we could use that as a prompt:
+
+Here's what we settled on:
+
+```javascript
+messages: [
+    {role: "system", content: "always respond with content for a Node-RED function node, and don't add any commentary, always use const or let instead of var. Always return msg, unless told otherwise."},
+    {role: "user", content: prompt}
+],
+```
+
+Here we send a `system` prompt in order to setup ChatGPT, and then follow that immediately with whatever the user has typed.
+From our (limited) testing, this has given us fairly reliable results.
+
+Breaking this prompt down:
+
+- ***"Always respond with content for a Node-RED function node"***: Ensured no surrounding `function () {}` definition and set expectations that the function would deal with a `msg` and likely `msg.payload` object.
+- ***"Don't add any commentary"***: ChatGPT likes to, well, chat. It would always return raw text justifying decisions, etc. Here, we just wanted the code.
+- ***"Always use const or let instead of var"***: This was Steve being picky.
+- ***"Always return msg, unless told otherwise"***: We found this wasn't mostly required, but occassitional it would try to return a different variable, and we'd lose context of `msg.payload`, or other data stored in `msg`. So this just made sure we had the consistency.
```

Comment:
```suggestion
- ***"Always return msg, unless told otherwise"***: We found this wasn't mostly required, but occasionally it would try to return a different variable, and we'd lose context of `msg.payload`, or other data stored in `msg`. So this just made sure we had the consistency.
```